File: ts-attempt2/src/transaction.ts
// src/transaction.ts

export class Transaction {
  private _institution: string;
  private _date: string;
  private _payee: string;
  private _amount: string;
  private _category: string;

  constructor(institution: string, date: string, payee: string, amount: string, category: string) {
    this._institution = institution;
    this._date = this.validateDate(date);
    this._payee = payee;
    this._amount = this.validateAmount(amount);
    this._category = category;
  }

  static createTransaction(institution: string, date: string, payee: string, amount: string, category: string): Transaction {
    if (category === 'SPLIT') {
      return new TransactionWithSplits(institution, date, payee, amount, category);
    }
    return new Transaction(institution, date, payee, amount, category);
  }

  private validateDate(date: string): string {
    if (!/\d{4}-\d{2}-\d{2}/.test(date)) {
      throw new Error('Invalid date format');
    }
    return date;
  }

  private validateAmount(amount: string): string {
    if (isNaN(parseFloat(amount))) {
      throw new Error('Invalid amount format');
    }
    return amount;
  }

  get institution(): string {
    return this._institution;
  }

  get date(): string {
    return this._date;
  }

  get payee(): string {
    return this._payee;
  }

  get amount(): string {
    return this._amount;
  }

  get category(): string {
    return this._category;
  }

  toCsv(headers: string[], values: string[]): string {
    return values.join(',');
  }
}

export class TransactionWithSplits extends Transaction {
  private _splits: TransactionSplit[];

  constructor(institution: string, date: string, payee: string, amount: string, category: string) {
    super(institution, date, payee, amount, category);
    this._splits = [];
  }

  get splits(): TransactionSplit[] {
    return this._splits;
  }

  getSplits(): TransactionSplit[] {
    return this._splits;
  }
}

export class TransactionSplit extends Transaction {
  private _parent: TransactionWithSplits;

  constructor(parentTransaction: TransactionWithSplits, institution: string, date: string, payee: string, amount: string, category: string) {
    super(institution, date, payee, amount, category);
    this._parent = parentTransaction;
    parentTransaction.splits.push(this);
  }

  get parent(): TransactionWithSplits {
    return this._parent;
  }
}

export default {
  Transaction,
  TransactionWithSplits,
  TransactionSplit
};

File: py-attempt/src/Transaction.py
#!/usr/bin/env python3
# transaction.py

from datetime import datetime

class Transaction:
    def __init__(self, institution, date, payee, amount, category):
        self._institution = institution
        self._date = self._validate_date(date)
        self._payee = payee
        self._amount = self._validate_amount(amount)
        self._category = category

    @property
    def institution(self):
        return self._institution

    @property
    def date(self):
        return self._date

    @property
    def payee(self):
        return self._payee

    @property
    def amount(self):
        return self._amount

    @property
    def category(self):
        return self._category

    @staticmethod
    def _validate_date(date_str):
        try:
            datetime.strptime(date_str, "%Y-%m-%d")
            return date_str
        except ValueError:
            raise ValueError("Invalid date format")

    @staticmethod
    def _validate_amount(amount_str):
        try:
            float(amount_str)
            return amount_str
        except ValueError:
            raise ValueError("Invalid amount format")

    @classmethod
    def create_transaction(cls, institution, date, payee, amount, category):
        if category == "SPLIT":
            return TransactionWithSplits(institution, date, payee, amount, category)
        return cls(institution, date, payee, amount, category)

    def to_csv(self, headers, values):
        return ",".join(values)

class TransactionWithSplits(Transaction):
    def __init__(self, institution, date, payee, amount, category):
        super().__init__(institution, date, payee, amount, category)
        self._splits = []

    @property
    def splits(self):
        return self._splits

    def get_splits(self):
        return self._splits

class TransactionSplit(Transaction):
    def __init__(self, parent_transaction, institution, date, payee, amount, category):
        super().__init__(institution, date, payee, amount, category)
        self._parent = parent_transaction
        parent_transaction.splits.append(self)

    @property
    def parent(self):
        return self._parent
